### ISSUE TITLE
The class weights implementation is incorrect

### ISSUE THREAD

**Issue by borundev** [pytorch/pytorch#61309]
## 🐛 Bug

The class weight implementation on a minibatch multiplies each data points loss with the corresponding weight but then divides by the sum of the weights. This means that the contribution of the same data point to the overall loss is dependent on the other members in the batch which is not the right behavior. This issue is explained in more detail in 

https://towardsdatascience.com/class-weights-for-categorical-loss-1a4c79818c2d

## To Reproduce

1. Create some fake fake data, labels and weights
2. Compute F.cross_entropy(input_t,target_t,weight=weight_t)
3. Compute (-torch.nn.LogSoftmax(1)(input_t)[range(3),target_t]*weight[target_t]).sum()/weight[target_t].sum()

https://colab.research.google.com/gist/borundev/5440ad77464d01672cda7dff06d89ebc/class-weights.ipynb

## Expected behavior

I would expect the denominator in the expression to be the number of rows in the minibatch.

## Environment

PyTorch version: 1.9.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0
Libc version: glibc-2.26

Python version: 3.7.10 (default, May  3 2021, 02:48:31)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: False
CUDA runtime version: 11.0.221
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.9.0+cu102
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.10.0
[pip3] torchvision==0.10.0+cu102
[conda] Could not collect

**Comment by heitorschueroff** [comment]
Marking high priority to review in triage meeting for potential incorrect implementation

**Comment by anjali411** [comment]
cc. @albanD

**Comment by albanD** [comment]
From the offline discussion we just had, both approach have different nice properties and drawbacks.

In particular, the proposed approach allow values to be independent of other elements in the batch.
But the rescaling by the weights ensures that the loss value for the important sample (assuming some weight are significantly larger than others) remains in a reasonable range, while samples with smaller weights gets much smaller values. This ensure numerical stability for the important sample during the backward pass.

Thanks for sharing the blog post as well, but I am not sure I follow the argument there.
In particular, it mentions that if there are two elements in the batch and the other element has a very different weight, then it will modify the gradient of the first sample. I think this is exactly the expected behavior actually.
Keeping in mind that the final loss is actually the sum (or mean) of all the samples in the dataset, we don't expect minibatches (after reduction) to have very different values as that would mean you are doing a pretty bad job at approximating the true loss.
So in that case, the rescaling by the weights actually does a better job at approximating the true loss.

**Comment by borundev** [comment]
I don't quite agree with the following statement "the rescaling by the weights ensures that the loss value for the important sample (assuming some weight are significantly larger than others) remains in a reasonable range, while samples with smaller weights gets much smaller values.".  I am assuming that you are open to the possibility of assigning weights not just because of class imbalance but also because of some classes being harder to train. In such a case I can have a balanced dataset of cats and dogs but say a **weight of 5 for cats and 1 for dogs**. Then **in a batch of all cats** (which is possible for small batch sizes) the **'effective weight' of cats will be 1 not 5**. Similarly, **** So in this case you can see that the effect of weights completely disappeared. These are extreme but not unlikely cases for small batch sizes.  In the other intermediate cases you also see such undesired effects.

What's particularly worrisome is that the above behavior is not expected by a user. Indeed I was working on a problem where I had to have small batches and the above **effectively leads to a fluctuating learning rate which messes with learning rate schedulers**.

**Comment by zou3519** [comment]
As a data point, @rwightman's https://github.com/rwightman/efficientnet-jax/blob/a65811fbf63cb90b9ad0724792040ce93b749303/jeffnet/common/loss.py#L31-L52 implements a weighted cross entropy by multiplying the unreduced loss by weight and then reducing with a mean()

**Comment by jbschlosser** [comment]
@borundev Thanks for the additional insight! I can certainly see how the current behavior is not great when a class is harder to train in general and the weight is desired independent of other items within the batch. That said, is it possible that the current PyTorch behavior could be preferred for the class imbalance case?

To cover this case as well as avoid breaking BC, I propose we add a new reduction mode that implements the TF-like semantics. That is, leave `reduction='mean'` as-is but add `reduction='truemean'` (name TBD) that divides by N unconditionally.

**Comment by borundev** [comment]
@jbschlosser I don't think the current behavior is preferred for the imbalanced class case either. I have the analysis in the linked colab notebook but I will add it here for the discussion. So basically we want to look at the stochastics of 

![image](https://user-images.githubusercontent.com/8229727/126897135-10ae91c7-dae3-48a1-825d-4c0fbdeb793f.png)

and for the probability we take the class probabilities and for the weights we take the usual weights one take for imbalanced classes with probability p. So we want to compute the mean and standard deviation of the above quantity 

![image](https://user-images.githubusercontent.com/8229727/126897170-ae543b0e-ba6f-4640-9dfb-1b257018d597.png)

Of course there is no need to compute it analytically and the plots for various batch sizes are given below

![image](https://user-images.githubusercontent.com/8229727/126897198-dd387972-ee9f-4b79-a487-3e3bb9bcc903.png)

Note the y-axis is not visible inline but if you click on the image it should be visible.

We see that with the PyTorch implementation the effective learning rate from .75 to 2 times the learning rate the user provided. I don't think this is the right behavior. Any issues with numerical stability @albanD pointed out should be taken care at the level of learning rate itself and not come through this implementation of class weights.

**Comment by albanD** [comment]
Hi,

Thanks for all the details!

What Joel was mentioning above is that, beyond the question of which one we want to have as a default, there is some backward compatibility constraints that we have.
We cannot just change the default as it would potentially change a lot of user-code in quite un-expected ways.
So we would need to go through the whole cycle to avoid any silent breakage for the users:
- Adding the new reduction as an opt-in and deprecate the current default, asking users to specify a reduction explicitly
- Remove the default altogether, forcing the users to always specify the reduction
- Add a new default that is the new behavior that we want
So that will take 3+ releases before it happens and will be quite a bit of work.

**Comment by jbschlosser** [comment]
Note that the current PyTorch implementation for class weights is also problematic when expanding `CrossEntropyLoss` to support "soft" labels (target class probabilities); attempted in #61044.

Users would reasonably expect that, for the same inputs:
1. Weighted CE computed with soft, one-hot labels == weighted CE computed with hard labels
2. CE computed with `weight=None` == CE computed with weights of all ones

Unfortunately, if we generalize the PyTorch implementation for class weights to the soft label case, this requires weighting by both target probabilities and class weights, which breaks 2 (see https://github.com/pytorch/pytorch/pull/61044#discussion_r675811766).

**Comment by idc9** [comment]
@jbschlosser perhaps calling it `reduction=`unweighted_mean`' would be more informative? If I saw `truemean` I would be confused/think something fancy was going on.

Another reason to prefer unweighted mean reduction is that (I imagine) people expect the mini-batch mean to be an unbiased estimated of the full gradient*, which the weighted mean reduction breaks. 

*ok we are randomly shuffling the data, not randomly subsampling, but my understanding is that the results of random shuffling should mimic true random subsampling

**Comment by jbschlosser** [comment]
> @jbschlosser perhaps calling it `reduction='unweighted_mean'` would be more informative? If I saw `'truemean'` I would be confused/think something fancy was going on.

Definitely agree that `reduction='truemean'` is not a great name. However, I think `reduction='unweighted_mean'` might be misleading too - that indicates to me that `weight` is left out entirely from the mean calculation. I think what we want is still technically a weighted mean calculation, just with an unweighted (correct) denominator of N.

Note that [`KLDivLoss`](https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html) uses `reduction='batchmean'`, although the problem being addressed is different.

Once we nail a name down, we'd accept a PR providing provide this reduction mode, as it's difficult to calculate manually in certain cases (e.g. if `ignore_index` is used).

**Comment by idc9** [comment]
ah good point about `unweighted_mean`.  `size_average`/`size_mean` would have been good but those are already used for something else.  

`size_normalized_mean` could be an option. `weight_normalized_mean` could then be an option for the current version (if you ever were to rename it). Perhaps this is too verbose though.

`batchmean`/`batch_mean`  seems like a good option to me

**Comment by rehno-lindeque** [comment]
Came here to suggest splitting `'mean'` into:

* `'ordinary_mean'`: After all it is an ordinary mean of the weighted sum. The weighted sum is coincidental.
* `'weight_normalized_mean'`: It is just the weighted mean, but as folks before me pointed out, that is still confusing since all the other reductions include a weighted sum.

Leaving the `reduction='mean'` would continue to confuse people into perpetuity, since most people would look at that and immediately conclude that it's just a convenience for `Tensor.mean()`. Finding out that the assumption is wrong after debugging or combing over documentation is not a great experience.

**Comment by nathanfranke** [comment]
I don't know how related this is. But the issue title still follows my understanding so far.

When respective target is zero (e.g. `target = [0, 1]`, `weight = [w, 1]`), weights have no effect, when respective target is onehot (e.g. `target = [1, 0]`, `weight = [w, 1]`), weights only act as a constant multiplier, 1/w, which shouldn't affect learning other than perhaps changing learn rate.

My overall goal: Weigh incorrect predictions more than correct predictions. Descend from confident incorrect faster than to confident correct. The prediction (output) should quickly descend to [0.5, 0.5], then slowly nudge towards better predictions according to weights as "confidence" factors.

![image](https://github.com/pytorch/pytorch/assets/14253836/408956e6-5155-4ad5-9cd1-bf7dd2ecbfcb)

<details>
<summary>Code</summary>
<br>

```py
import torch

print("Cost; target = [0 1]")
print("columns", end="")
for o in [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]:
    print(f", o = [{o:.1f} {1-o:.1f}]", end="")
print()
for w in [0.02, 0.1, 1.0, 10.0, 50.0]:
    print(f"w = [{w:.2f} 1.0]", end="")
    target = torch.tensor([0.0, 1.0])
    weight = torch.tensor([w, 1.0])
    for o in [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]:
        outputs = torch.tensor([o, 1.0 - o])
        criterion = torch.nn.CrossEntropyLoss(weight=weight)
        cost = criterion(outputs.unsqueeze(0), target.unsqueeze(0))
        print(f", {cost.item():.5f}", end="")
    print()
print()

print("Cost; target = [1 0]; values divided by w")
print("columns", end="")
for o in [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]:
    print(f", o = [{o:.1f} {1-o:.1f}]", end="")
print()
for w in [0.02, 0.1, 1.0, 10.0, 50.0]:
    print(f"w = [{w:.2f} 1.0]", end="")
    target = torch.tensor([1.0, 0.0])
    weight = torch.tensor([w, 1.0])
    for o in [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]:
        outputs = torch.tensor([o, 1.0 - o])
        criterion = torch.nn.CrossEntropyLoss(weight=weight)
        cost = criterion(outputs.unsqueeze(0), target.unsqueeze(0)) / w
        print(f", {cost.item():.5f}", end="")
    print()
print()

```
</details>

<details>
<summary>Output</summary>
<br>

```
Cost; target = [0 1]
columns, o = [0.0 1.0], o = [0.2 0.8], o = [0.4 0.6], o = [0.6 0.4], o = [0.8 0.2], o = [1.0 0.0]
w = [0.02 1.0], 0.31326, 0.43749, 0.59814, 0.79814, 1.03749, 1.31326
w = [0.10 1.0], 0.31326, 0.43749, 0.59814, 0.79814, 1.03749, 1.31326
w = [1.00 1.0], 0.31326, 0.43749, 0.59814, 0.79814, 1.03749, 1.31326
w = [10.00 1.0], 0.31326, 0.43749, 0.59814, 0.79814, 1.03749, 1.31326
w = [50.00 1.0], 0.31326, 0.43749, 0.59814, 0.79814, 1.03749, 1.31326

Cost; target = [1 0]; values divided by w
columns, o = [0.0 1.0], o = [0.2 0.8], o = [0.4 0.6], o = [0.6 0.4], o = [0.8 0.2], o = [1.0 0.0]
w = [0.02 1.0], 1.31326, 1.03749, 0.79814, 0.59814, 0.43749, 0.31326
w = [0.10 1.0], 1.31326, 1.03749, 0.79814, 0.59814, 0.43749, 0.31326
w = [1.00 1.0], 1.31326, 1.03749, 0.79814, 0.59814, 0.43749, 0.31326
w = [10.00 1.0], 1.31326, 1.03749, 0.79814, 0.59814, 0.43749, 0.31326
w = [50.00 1.0], 1.31326, 1.03749, 0.79814, 0.59814, 0.43749, 0.31326
```
</details>

Of course, thank you, community and contributors

**Comment by jneeven** [comment]
I fully agree with @rehno-lindeque:
> Leaving the `reduction='mean'` would continue to confuse people into perpetuity, since most people would look at that and immediately conclude that it's just a convenience for `Tensor.mean()`. Finding out that the assumption is wrong after debugging or combing over documentation is not a great experience.

I've had to debug for several hours before it occurred to me that this discrepancy might somehow be intentional. Especially because `reduction="sum"` does work as expected... Please clarify the documentation to reflect that it is not taking the ordinary mean 😢

**Comment by ezyang** [comment]
I will happily help review a doc fix. But I'm also going to bump this to our triage review meeting about if there's something we can do about this.

**Comment by janeyx99** [comment]
From triage discussion: two things we can do
(1) @mikaylagawarecki proposed that we should better document how mean works and
(2) @jbschlosser suggested we should intro a new reduction mode that is more intuitive

Will leave it up to them to further comment
